packages/horizon/contracts/staking/HorizonStakingExtension.sol

```diff
@@ -255,6 +263,92 @@ contract HorizonStakingExtension is HorizonStakingBase, IHorizonStakingExtension
         return __DEPRECATED_thawingPeriod;
     }
 
+    function legacySlash(
+        address indexer,
+        uint256 tokens,
+        uint256 reward,
+        address beneficiary
+    ) external override onlySlasher notPaused {
+        ServiceProviderInternal storage indexerStake = _serviceProviders[indexer];
+
+        // Only able to slash a non-zero number of tokens
+        require(tokens > 0, "!tokens");
+
+        // Rewards comes from tokens slashed balance
+        require(tokens >= reward, "rewards>slash");
+
+        // Cannot slash stake of an indexer without any or enough stake
+        require(indexerStake.tokensStaked > 0, "!stake");
+        require(tokens <= indexerStake.tokensStaked, "slash>stake");
+
+        // Validate beneficiary of slashed tokens
+        require(beneficiary != address(0), "!beneficiary");
+
+        // Slashing more tokens than freely available (over allocation condition)
+        // Unlock locked tokens to avoid the indexer to withdraw them
+        uint256 tokensAvailable = indexerStake.tokensStaked -
+            indexerStake.__DEPRECATED_tokensAllocated -
+            indexerStake.__DEPRECATED_tokensLocked;
+        if (tokens > tokensAvailable && indexerStake.__DEPRECATED_tokensLocked > 0) {
+            uint256 tokensOverAllocated = tokens - tokensAvailable;
+            uint256 tokensToUnlock = MathUtils.min(tokensOverAllocated, indexerStake.__DEPRECATED_tokensLocked);
+            indexerStake.__DEPRECATED_tokensLocked = indexerStake.__DEPRECATED_tokensLocked - tokensToUnlock;
+            if (indexerStake.__DEPRECATED_tokensLocked == 0) {
+                indexerStake.__DEPRECATED_tokensLockedUntil = 0;
+            }
+        }
+
+        // Remove tokens to slash from the stake
+        indexerStake.tokensStaked = indexerStake.tokensStaked - tokens;
+
+        // -- Interactions --
+
+        // Set apart the reward for the beneficiary and burn remaining slashed stake
+        _graphToken().burnTokens(tokens - reward);
+
+        // Give the beneficiary a reward for slashing
+        _graphToken().pushTokens(beneficiary, reward);
+
+        emit StakeSlashed(indexer, tokens, reward, beneficiary);
+    }
+
+    /**
+     * @notice Withdraw undelegated tokens once the unbonding period has passed.
+     * @param indexer Withdraw available tokens delegated to indexer
+     */
+    function legacyWithdrawDelegated(
```

Comment:
Maybe including this in withdrawDelegated() (as suggested in the audit report) would make it simpler for Explorer? otherwise Explorer might need to do some tricky handling to figure out what function to call.

Reply:
I think the explorer will still need to do some handling to know which type of withdraw it needs to use, right? Once Horizon is deployed undelegating will use the new thaw requests so it'll need to call the new withdraw. But also now that I think about it we won't be able to remove this functionality since we'll still need to support withdrawing after transition period I guess so I'll move it to `withdrawDelegated()`.

Reply:
Moved. Do you mind reviewing again? Thanks.